src/System.Management.Automation/engine/hostifaces/LocalConnection.cs

```diff
@@ -867,19 +838,21 @@ private void DoCloseHelper()
                 }
 
                 if ((hostRunspace == null) || (this == hostRunspace))
-                {                   
+                {
+                    bool stopTranscription = true;
                     // We should close transcripting only if we are closing the last opened runspace.
                     foreach (Runspace runspace in RunspaceList)
                     {
                         // At this stage, the last opened runspace should be at closing state.
                         if (runspace.RunspaceStateInfo.State == RunspaceState.Opened)
                         {
-                            return;
+                            stopTranscription = false;
+                            break;
                         }
                     }
 
                     PSHostUserInterface host = executionContext.EngineHostInterface.UI;
-                    if (host != null)
+                    if (host != null && stopTranscription)
```

Comment:
Sorry if you already explained this change in some previous conversations, but I didn't find any.
The original code seems already makes sure that `StopAllTranscribing` will be called only if all runspaces are at closing state. As for `AmsiUtils.Uninitialize()`, it looks to me doesn't need to be called for every Runspace close operation, but stead should only be called once when the last Runspace is closed.

Reply:
@lzybkr had a question about this section in that previously it would return out of this method early 
 if it's not the last runspace so that transcription would not be stopped.  Because it returned early, the rest of the closing actions would not be executed.

Reply:
The rest of the closing method is `AmsiUtils.Uninitialize();` which I think should only execute once when the last Runspace is closing. It looks to me `AmsiUtils.Init()` should run only once and then can be shared across all Runspaces.

Reply:
I remember `AMSI` is thread safe and @PaulHigin even fixed a bug previously to remove the locks when calling `AmsiUtils.ScanContent`. So I think it should be initialized once and used across all Runspaces.

Reply:
I'll change it so if there's no open runspaces left, transcription is stopped and AmsiUtils.Uninitialize() is called.